### ISSUE TITLE
Support graceful provider release deprecation

### ISSUE THREAD

**Issue by t0yv0** [pulumi/pulumi#12155]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Consider adding support in the Pulumi CLI for the scenario of deprecated provider releases. This scenario arises when a critical bug is discovered in a provider release and the maintainers want to encourage all he users to immediately upgrade onto a newer version that contains a fix, as well as reduce the number of people installing the affected version and thus localize the blast radius.

The issue is complicated because of a multi-language nature of Pulumi solutions. Deprecating a release may involve dealing with GitHub and S3 releases for the provider, as well as NPM, Maven, NuGet, etc SDKs. While some of the package ecosystem support this scenario (num-deprecate), others require further work.

In light of this it would be very helpful if provider maintainers could quickly flag a release as deprecated and have Pulumi CLI issue a graceful message encouraging an upgrade during auto-install of the provider. It is probably too much to do additional net calls to lookup deprecation status every time the provider is used, but doing it just at auto-install phase already seems to cover a lot of ground and would be very helpful.

Our current practice for the deprecated release situation is deleting the GitHub release and provider binaries. This leads to unhelpful error messages. Users trying to install the affected version get a 404-style merge. More seriously, users that have already installed the affected version may encounter errors that interrupt updates and require manual stack state clean up via `pulumi stack export` and `pulumi stack import`).

See also:

- https://github.com/pulumi/pulumi-random/issues/258
- https://github.com/pulumi/pulumi/issues/12185
- https://github.com/pulumi/ci-mgmt/issues/336
- https://github.com/pulumi/pulumi-command/issues/183
- [Postmortem](https://docs.google.com/document/d/1PMQEX_wMsJqUwbA-98e37BAIPbAgZ2zrqKeygADo4lc/edit?usp=sharing)

**Comment by Zaid-Ajaj** [comment]
> It is probably too much to do additional net calls to lookup deprecation status every time the provider is used, but doing it just at auto-install phase already seems to cover a lot of ground and would be very helpful.

That seems reasonable. For example, instead of deleting a release, we could manually add a `deprecated.txt` file to the release assets and look that up during the auto-install phase. The contents of the asset could provide additional context as to why the provider has been deprecated and the CLI could show that to users.  

This solution is simple but does feel a bit _ad hoc_, happy to hear about alternatives. 

> More seriously, users that have already installed the affected version may encounter errors that interrupt updates and require manual stack state clean up via pulumi stack export and pulumi stack import).

If we detect that a version is deprecated, I feel like we should do clean up automatically for the users.

**Comment by t0yv0** [comment]
Coming back to this with the AWS v6.0.0 deprecation attempt.

The manual steps are better documented in a playbook so while not perfect this currently is acceptable to keep using a playbook instead of automation for this rare event https://github.com/pulumi/platform-providers-team/blob/main/playbooks/retracting-provider-releases.md

However one scenario is really still not having any good answers and it is https://github.com/pulumi/pulumi-aws/issues/2695 - this is currently affecting one of our enterprise customers and requires a manual workaround. 

The issue is that references to the deprecated release can make it into Pulumi Stacks, and the engine tries to download that release of the provider, even for replacing or destroying the stack. 

We need some story here to be able to instruct the engine to redirect gracefully to use a supported version instead of failing with a "cannot download" error.

**Comment by Frassle** [comment]
> We need some story here to be able to instruct the engine to redirect gracefully to use a supported version instead of failing with a "cannot download" error.

This should work today if you run an `up` with no changes. The engine will warn it couldn't download it but shouldn't error.

**Comment by Frassle** [comment]
But a 'state replace provider' command might be useful for this and some other edgecase scenarios as well.

**Comment by t0yv0** [comment]
Well the 'up' with no changes does work. We'll try to get back a better repro on where the customer's at.

**Comment by t0yv0** [comment]
Yeah. `state replace provider` could be useful. Ideally there's something we can do that requires no user action at all, but having a clean warning with instructions is also good. This is not a very common occurrence.

**Comment by Frassle** [comment]
> Ideally there's something we can do that requires no user action at all, 

Well I think the ideal for that would some channel that the CLI can query for if a version is "bad" and if it is then use the new version even for deletes such that the bad version isn't used at all.

**Comment by t0yv0** [comment]
Yes precisely. Or even a redirect. IF we could delete 6.0.0 release and publish a "use 5.42.1" instead kind of instruction.
